### DISCUSSION TITLE
"out of order read" error when trying to write sequential file from threads

### DISCUSSION THREAD

**Discussion by richardmonette** [libvips/libvips#2923] [General]
My scenario is that I've got an image opened with VIPS_ACCESS_SEQUENTIAL, am performing some transformations such as rotation and then I'd like to, in parallel, with threads, save it to WebP and JPEG. 

I am running into errors such as "could not write libvips image to byte buffer: vips_image_get: field \"orientation\" not found\nvips_image_get: field \"orientation\" not found\nvips_image_get: field \"jpeg-chroma-subsample\" not found\nvips_image_get: field \"jpeg-chroma-subsample\" not found\nVipsJpeg: out of order"

This would seem to sort of make sense, if the sequential buffer is expecting the reads to come in order, then two threads reading at different points in the buffer would potentially be "out of order."

What I am hoping to do is somehow flush the operation queue such that I have an image I can pass to multiple write functions, or perhaps make copies which do the same, and each of which can be accessed without triggering the out of order errors. Perhaps something like https://www.libvips.org/API/current/VipsImage.html#vips-image-new-from-memory-copy?

**Comment by jcupitt** [comment]
Hi @richardmonette,

You can only use sequential images once, in a single top to bottom read. Did you see the section in the docs on how they work? I don't remember if we talked about this:

https://www.libvips.org/API/current/How-it-opens-files.html

You'll need to open them a second time if you want to use them again.

**Comment by richardmonette** [comment]
Hey John, thanks for your message. 

Yep, I read through that section - I was wondering though, do I have to fully load the images again, or is there some way I could copy them, etc. _after_ I've done the transformation operations? What I am trying to avoid is having to run my entire transform pipeline from scratch on each image, and instead do everything which is the same anyways, but then have the final `vips_jpegsave_buffer` and `vips_webpsave_buffer` run in parallel.

Question as sort of a related aside; from what I understand, the image operations will run in the number of threads available to libvips, is that already the same for the save buffer operations? (or are they sinks? https://github.com/libvips/libvips/blob/master/doc/How-it-works.md)

basically what I want is:

```
load -> transform -> write_webp
                | -> write_jpeg
``` 

where the writing is in parallel. I'm able to achieve this with sequential disabled, but because of https://github.com/libvips/libvips/issues/2739 I have to keep it on.

I'd like to avoid having to do

```
load -> transform -> write_webp
load -> transform -> write_jpeg
```

where I am having to fully load the image several times and running most of the transform steps as duplicated effort.

**Reply by jcupitt** [reply]
Ah I see. You'd need to do `copy_memory()` between the transform and save, so:

```
load -> transform -> copy_memory -> write_webp
                               | -> write_jpeg
```

You can run the saves at the same time. You can leave seq on for the inputs.

**Reply by richardmonette** [reply]
Right, makes sense. To be clear, are you suggesting the method `vips_image_copy_memory ()` https://www.libvips.org/API/current/VipsImage.html#vips-image-copy-memory? 

There seems to be a few methods which do various types of copying, and I just want to make sure that's the one you recommend (and that it will solve the sequential access issues.)

**Reply by richardmonette** [reply]
Quick follow up on this, perhaps I am just using `vips_image_copy_memory` wrong but I am running into segmentation violations trying to write the copied images, e.g.

```
[signal SIGSEGV: segmentation violation code=0x2 addr=0x128 pc=0x1046837f4]

runtime stack:
runtime.throw({0x10342b46e?, 0x16d6bad90?})
	/Users/richardmonette/.dev/go/1.18/src/runtime/panic.go:992 +0x50
runtime.sigpanic()
	/Users/richardmonette/.dev/go/1.18/src/runtime/signal_unix.go:802 +0x1e8

goroutine 13 [syscall]:
runtime.cgocall(0x1033f17f4, 0x14000665c88)
	/Users/richardmonette/.dev/go/1.18/src/runtime/cgocall.go:157 +0x54 fp=0x14000665c50 sp=0x14000665c10 pc=0x102a8e1a4
github.com/pkg/libvips._Cfunc_lv_write_webp
```

**Reply by richardmonette** [reply]
Apologies for the repeated messages, figured I'd add a bit more info from my debugging efforts.

Noticed a couple things:

1) If I do the `vips_image_copy_memory` but only have one thread, then it does seems OK, so the segmentation violation does seem to come from some overlap between the threads

2) When the error occurs if I trap using 

```
if (!(out = vips_image_copy_memory(in)) ) {
        vips_error_exit("vips_image_copy_memory appears to be failing");
    }
```

then I see

```
(process:50346): VIPS-WARNING **: 14:49:17.320: error in tile 0 x 0

(process:50346): VIPS-WARNING **: 14:49:17.320: error in tile 0 x 8
libvips_go: vips_image_copy_memory appears to be failing
vips_image_get: field "orientation" not found
vips_image_get: field "orientation" not found
vips_image_get: field "jpeg-chroma-subsample" not found
VipsJpeg: out of order read at line 3252

(process:50346): GLib-CRITICAL **: 14:49:17.321: g_hash_table_lookup: assertion 'hash_table != NULL' failed

(process:50346): GLib-CRITICAL **: 14:49:17.321: g_hash_table_lookup: assertion 'hash_table != NULL' failed
```

which suggests I am still running into the `out of order read`, despite the usage of `vips_image_copy_memory`

**Reply by jcupitt** [reply]
I'm not sure I understand what you're running in parallel. You need to run the first bit in one thread (up to and including the `copy_memory`),  then you can run the two saves from two different threads, but both reading the memory result.

Could you make a complete C, C++, Python or Ruby program that shows this issue?

**Reply by richardmonette** [reply]
Hey John, I've put a complete C example below (after these notes) which is similar to what I am trying to do and causes seg faults. It's a bit different from what I have running in Go + C, but maybe if we could get a working C example, then I could translate it back to my other code.

The thing I noticed was that running `vips_image_copy_memory` once is OK, but the second time when I check

```
VipsImage *copiedImage = vips_image_copy_memory(in);
if (!copiedImage) {
     vips_error_exit("vips_image_copy_memory appears to be failing");
}
```

the copy is failing, with the "out of order read" message in the exit errors. I am wondering if that is because, as the docs say "If image is already a memory buffer, just ref and return." maybe what's happening there is because I am copying an existing buffer it's just making a reference to it, and not really copying it in a manner where the SEQUENTIAL guard is being removed?

I've been experimenting with `vips_image_new_from_memory_copy`, which seems to allow my program to proceed, although I think I have parameters not quite right because I am seeing images such as

<img width="408" alt="Screen Shot 2022-07-18 at 11 49 08 AM" src="https://user-images.githubusercontent.com/1041278/179550943-5b336d60-1b58-4b5f-8637-9049495b1b27.png">

Is there an easy way to get the values for this function from the source image?

Here is the C code:

```
#include <stdio.h>
#include <vips/vips.h>
#include <pthread.h>

// gcc -g -Wall vips_image_copy_test.c `pkg-config vips --cflags --libs`

struct ThreadArgs {
  VipsImage *in;
  int thread_id;
}; 

void *writeImage(void *vargp)
{
    struct ThreadArgs *ta = (struct ThreadArgs *)vargp;

    printf("writeImage %d... \n", ta->thread_id);

    char str[80];

    if (ta->thread_id == 1) {
      sprintf(str, "writeImage%d.jpg", ta->thread_id);
      vips_jpegsave(ta->in, str, NULL);
    } else if (ta->thread_id == 2) {
      sprintf(str, "writeImage%d.webp", ta->thread_id);
      vips_webpsave(ta->in, str, NULL);
    }

    return NULL;
}

int main( int argc, char **argv )
{
  VipsImage *in;

  if( VIPS_INIT( argv[0] ) )
    vips_error_exit( NULL ); 
  
  if( !(in = vips_image_new_from_file( "test.jpg", "access", VIPS_ACCESS_SEQUENTIAL, NULL )) )
    vips_error_exit( NULL );

  VipsImage *out;
  vips_resize(in, &out, 2.0, NULL);

  printf("Create thread 1\n");

  VipsImage *copy1 = vips_image_copy_memory(out);

  struct ThreadArgs *t1 = (struct ThreadArgs *)malloc(sizeof(struct ThreadArgs));
  t1->in = copy1;
  t1->thread_id = 1;

  pthread_t pt1;

  pthread_create(&pt1, NULL, writeImage, (void *)t1);

  printf("Create thread 2\n");

  VipsImage *copy2 = vips_image_copy_memory(out);

  struct ThreadArgs *t2 = (struct ThreadArgs *)malloc(sizeof(struct ThreadArgs));
  t2->in = copy2;
  t2->thread_id = 2;

  pthread_t pt2;

  pthread_create(&pt2, NULL, writeImage, (void *)t2);

  printf("Join threads\n");
  
  pthread_join(pt1, NULL);
  pthread_join(pt2, NULL);

  printf("Done\n");

  return( 0 );
}
```

**Reply by jcupitt** [reply]
You have:

```
  VipsImage *copy1 = vips_image_copy_memory(out);

...

  VipsImage *copy2 = vips_image_copy_memory(out);
```

So you're reading `out` twice. 

Instead, just use `copy1` again.

**Reply by richardmonette** [reply]
Hey John, just wanted to follow up to say that looks like it will work. 

Thank you very much for your help with this! 

I'll write back again if we happen to find any follow on problems.

**Reply by richardmonette** [reply]
Hey John,

So it looks like, practically speaking, using the copy allows for the parallel encoding, but I'm getting some questions from my colleagues as to more theoretically why the copy allows for things to work. 

My hypothesis is that the copy forces libvips to evaluate the transform operations that have been otherwise deferred until that point, and then the resulting copied buffer is not marked for sequential access. Is this correct? If I am misunderstanding why the copy fixes the problem I'd be very keen to get your explanation as to why the fix works, so I could communicate that back to my team, such that they would be more comfortable with shipping this solution to production.

Again, thanks for taking the time to help me understand how libvips is operating in this scenario.

**Reply by jcupitt** [reply]
Yes, that's about right. It's really simple (or it seems that way to me after decades looking at it, heh).

libvips is a lazy, demand-driven library, although the API copies the more usual immediate mode pattern. When you do eg. (using python because it's concise, but the C/C++ APIs are the same underneath):

```python
a = b + c
```

It's not actually computing any pixels. `a` will have its header set up, so it looks as if there are pixels there (you can query width, height, etc.), but there's no pixel data. Instead, it'll create a new `add` pipeline stage, refer to this new stage as `a`, and this new node will join the `b` and `c` pipelines together with a callback to compute pixels.

Everything is like this: new stages are added to the pipeline, and nothing is ever calculated ... until you connect to a sink, such as file, memory or display output. When you run:

```
d = a.copy_memory()
```

It'll look at the `a` header to get width, height etc., allocate a huge chunk of memory, and then spawn a set of worker threads to compute those pixels. All the prior pipeline stages execute together and in parallel, and the memory slowly fills with data. If anything upstream of `a` or `b` is sequential, execution will respect that. Threads run horizontally, ie. each thread executes the entire pipeline.

Once computation is done, a new image `d` is created which wraps the memory area and is able to serve pixels from that. Now you can use that in any way you like, and no part of the `a` pipeline will execute, and so it won't be bound by any upstream restrictions.

Did you see the "how it works" page? It has a more technical introduction to libvips evaluation:

https://www.libvips.org/API/current/How-it-works.html

**Comment by richardmonette** [comment]
Hey John, follow up on this. I was noticing that the boost in performance I was expecting to see by running the writes in parallel hasn't quite materialized, and I was curious to share some data to get your perspective.

I wrote a tiny test program you can see here: https://gist.github.com/richardmonette/564e1a5b1f6453212a8d78d82cd50df8 which I used to get these numbers.

Running the two writes sequentially, I am seeing a total time of ~736ms, where ~130ms is writing the JPEG and ~604ms is writing the WebP. This seems straightforward, as the total time is just the two writes added together.

What's more interesting is that when running the writes in threads the total time is ~754ms, where 318ms is the JPEG and 754ms is the WebP. 

Some observations on the threaded scenario:

- The overall total time is similar, but slightly higher. Which is to say, using threads did not increase performance, if anything, it may have been slightly worse! 🤔 (I might have expected the wall clock time in the threaded scenario to be 604ms, since that's the longest work item)
- It looks like both the JPEG and WebP writes were slower from within the threads. So although the wall clock time was only slight worse, the total compute/effort time between the two threads was 1072ms, which is appreciably worse/more effort for the same result in the end.

I was wondering then, as I was expecting to see the wall clock time in the threading scenario being lower (i.e. only the duration of the longest write), am I doing something wrong in my code? 

Is there maybe some locking mechanism I am unaware of where only one of the writer libraries can actually do meaningful work at a time, so in practice I'm just waiting for some mutex to be handed back and forth?

Thanks for your continued support in this regard 🙂

**Reply by jcupitt** [reply]
I think there's some problem with your pthreads not running in parallel. I tried this:

```C
/* Compile with:
 *
 *      gcc -g -Wall parallel-encode.c `pkg-config vips --cflags --libs`
 *
 * RUn with:
 *
 *      mkdir x
 *      ./a.out ~/pics/k2.jpg 20 .jpg
 */

#include <vips/vips.h>

char *output_format = NULL;
VipsImage *image = NULL;

/* Count idle threads with this.
 */
VipsSemaphore idle_threads;

static void *
worker( void *data )
{
        char output_file[256];

        snprintf( output_file, 256, "x/tmp-%p%s", 
                g_thread_self(), output_format );
        if( vips_image_write_to_file( image, output_file, NULL ) )
                vips_error_exit( NULL );

        vips_semaphore_up( &idle_threads );

        return( NULL );
}

int
main( int argc, char **argv )
{
        int i;
        int n;
        GTimer *timer;
        double elapsed;

        if( VIPS_INIT( argv[0] ) )
                return( -1 );
        vips_cache_set_max( 0 );

        if( !(image = vips_image_new_from_file( argv[1], NULL )) )
                return( -1 );
        if( !(image = vips_image_copy_memory( image )) )
                return( -1 );

        n = atoi( argv[2] );

        output_format = argv[3];

        vips_semaphore_init( &idle_threads, n, "idle" );

        timer = g_timer_new();

        for( i = 0; i < n; i++ ) {
                vips_semaphore_down( &idle_threads );
                g_thread_new( NULL, (GThreadFunc) worker, NULL );
        }

        /* Wait for all threads to finish.
         */
        vips_semaphore_downn( &idle_threads, n );

        elapsed = g_timer_elapsed( timer, NULL );

        printf( "%.2g elapsed\n", elapsed );
        printf( "%.2g per save\n", elapsed / n );

        vips_shutdown();

        return( 0 );
}
```

I see:

```
$ ./a.out ~/pics/k2.jpg 1 .webp
0.23 elapsed
0.23 per save
$ ./a.out ~/pics/k2.jpg 10 .webp
0.29 elapsed
0.029 per save
```

So 10 webp saves in parallel is about 10x faster.
